sig-architecture/api-review-process.md

```diff
@@ -0,0 +1,245 @@
+
+    * Configuration files, flags, and command line arguments are all part of our user and script facing APIs and must be reviewed.
+
+    * Compiled-in APIs of the kube-apiserver
+
+    * Webhooks request/response formats in kube-apiserver
+
+    * HTTP APIs in kubelet
+
+    * plugins which are not covered by some other standards effort (e.g. CSI and CNI APIs would be deferred to those standards bodies)
+
+## End-states of Reviews
+
+The API review process can result in multiple outcomes depending on the content of the change. For example, a new API could be approved for being builtin, or rejected in favor of out-of-tree development. An API that is reviewed informationally, results in either suggestions for changes, or approval as-is. 
+
+## Information required from the submitter
+
+As much as possible, we will automate the detection of PRs that require API reviews.  Significant changes should be reviewed BEFORE they reach the PR stage, but this is the backstop for anything that gets missed.  Such automation may miss cases, so any PR can be flagged as "needs API review", which triggers this process.
+
+To begin the process, the proposer must produce:
+
+* Create an issue in the kubernetes-sigs/architecture-tracking repository that links to the relevant KEP or documentation
+
+* The KEPs/documentation should include A clear and thoroughly-researched justification on why the change or addition is needed, including, upgrade/downgrade considerations, and alternatives considered.
+
+* The proposer may follow one of two paths:
+
+    * Complete the coding of the API change.  Create a PR.  Request an API review on the PR.  (In future, the request will be automated based on detecting API changes). The API reviewer will /approve the PR, assuming the change was satisfactory
+
+    * Write the KEP/documentation including a detailed description of the API.  Request a review on that.  The API reviewer will note in the API review issue the commit SHA at which the KEP was reviewed.  The reviewee will note the API review issue number.  Later, when the PR is ready to add the API, the reviewee files another review ticket (per above path).  The review is expedited because the reviewer only needs to compare the current PR to what was previously approved.  This two step process allows for automation, and allows for KEPs to get API approval before moving forward.
+
+*  explicit approvals by SIG stakeholders are not checked in the API review process.  This is assumed to be given at the PR stage.
+
+* Any additional GitHub IDs associated with the submission (particularly in the case of multiple authors)
+
+* Acknowledgement that they have read and followed the existing API conventions document 
+
+## Information to be provided from the reviewer(s)
+
+If **approved**/**reviewed**:
+
+* Any non-blocking or nit suggestions should be documented in the review document that will be stored in the repository
```

Comment:
"the review document" is unclear - not defined.  Are we keeping a separate doc for API review from the code or KEP PR?  Should we require that API be cut-paste into a new PR *just* for API review?  Do we want to keep those forever?  Github's historical view is kind of useless for learning (because old state is obliterated).

Reply:
I had imagined the doc as something permanent in MD that would be iterated on between the reviewer(s) and submitter(s). If possible, I'd like to try that and see if it works, otherwise, I consider the other options you raise as viable alternatives.

Reply:
A code review of a .md file is no different from a code review of a .go file.  Github's history is not easy to follow.  Just to be clear,  you're proposing to create a new markdown file for every API change, which is NOT the actual code, and which we keep forever?  Why not just link to the PR containing the proposed changes?

Reply:
@thockin I think we can link in the tracking issue to the PR. It's just a little harder to follow after the fact IMO. But this process belongs to the reviewers/approvers, and we can tailor it as necessary over time.